### PR TITLE
Print panel improvements (Issue 2634)

### LINKED ISSUE #2634: Improve print panel
# Feature Request

**Is your feature request related to a problem? Please describe.**
We can make nice maps but ugly print layout. There is a low hanging fruit which would make the output much nicer.
Improvement ideas:
+ Style the map title
+ Add legend
+ Add optional imprint information
+ Add optional scale (either numeric or graphic, only useful on large scale)

**Describe the solution you'd like**
For now just improve the look and feel of the output.

**Describe alternatives you've considered**
For the future, some extra map-designer tool which would also allow to move and style the individual components of the map design, would be supercool.

**Additional context**
.

### PR DESCRIPTION
**Pull request by DailisLangovskis**
## Description

This PR introduces features, such as, adding styled title, optional map legend, optional imprint, optional map scale to an image, that can be printed or downloaded as .PNG image.

## Related issues or pull requests

closes #2634 

## Pull request type

Please check the type of change your PR introduces:

- [x] Bugfix
- [x] Feature
- [ ] Dependency updates
- [x] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe)
- [ ] I am unsure (we'll look into it together)

## Do you introduce a breaking change?

- [ ] Yes
- [x] No
- [ ] I am unsure (no worries, we'll find out)

## Checklist

- [x] I understand and agree that the changes in this PR will be licensed under the [MIT License]
- [x] I have followed the [guidelines for contributing](https://github.com/hslayers/hslayers-ng/blob/master/CONTRIBUTING.md)
- [x] The proposed change fits to the content of the [code of conduct](https://github.com/hslayers/hslayers-ng/blob/master/CODE_OF_CONDUCT.md)
- [x] I have added or updated tests and documentation, and the test suite passes (run `npm test` locally)
- [ ] I'm lost; why do I have to check so many boxes? Please help!

### CODE REVIEW

**Review by fzadrazil — CHANGES_REQUESTED** [review]
* It would be good to keep print settings when the Print panel is closed and opened again
* There is something broken when setting scalebar. Print button throws error then

```
core.mjs:6485 ERROR Error: Uncaught (in promise): TypeError: Cannot read properties of undefined (reading 'height')
TypeError: Cannot read properties of undefined (reading 'height')
    at HsPrintService._callee4$ (print.service.ts:109:62)
```

* When scalebar is turned on and set to "bar" type and "text" set to true, scale is not visible in preview

**Review by fzadrazil — APPROVED** [review]

**Comment by DailisLangovskis** [comment]
I will add tests after weekend, so it there is no rush, lets keep the PR for now

**Comment by jmacura** [comment]
I would change "Fill color" to "Text color" as it is more what it is.
obrazek

**Comment by jmacura** [comment]
Also, how hard would it be to add also "background color"? The text is often hardly readable on the map.

**Comment by DailisLangovskis** [comment]
Will do, and there won't be no problem adding background colour.

jmacura ***@***.***> (šajā datumā: piektd., 2022. g. 25.
febr. 23:06) rakstīja:

File: projects/hslayers/src/components/print/models/text-style.model.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import {xPos, yPos} from '../types/XY-positions.type';
+
+export type TextStyle = {
+  fillColor?: string;
+  strokeColor?: string;
```
Comment (jmacura) [inline projects/hslayers/src/components/print/models/text-style.model.ts:5]: What is this param good for? I can't see it used anywhere in the templates..

File: projects/hslayers/src/components/print/models/text-style.model.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  textSize?: string;
+  fontFamily?: string;
+  fontStyle?: string;
+  textDraw?: 'fill' | 'stroke';
```
Comment (jmacura) [inline projects/hslayers/src/components/print/models/text-style.model.ts:9]: Also, what is this param good for?

**Review by jmacura — COMMENTED** [review]
Code-wise comments:
- Why do you separate "types" and "models"? Is there some semantic difference? I would put everything into one folder ("models" probably...), if there is none.
- For the "utils" folder, I would probably call it just "constants", but that's a matter of personal taste for sure 🙂

**Comment by jmacura** [comment]
> Will do, and there won't be no problem adding background colour.

Great! To be clear, I mean this change for "Imprint style" and "Title style". Not for "Legend style", where the "Fill color" shall stay as it is (or changed to "Background color" for consistency). Thanks!

**Comment by DailisLangovskis** [comment]
> Code-wise comments:
> 
> * Why do you separate "types" and "models"? Is there some semantic difference? I would put everything into one folder ("models" probably...), if there is none.
> * For the "utils" folder, I would probably call it just "constants", but that's a matter of personal taste for sure 🙂

Agree, at first I ment them as interfaces, but after changes forgot to change them back. Will do.

File: projects/hslayers/src/components/print/models/text-style.model.ts (hunk 1 above)
Comment (DailisLangovskis) [reply projects/hslayers/src/components/print/models/text-style.model.ts:5]: StrokeColor I use for text stroke style.

File: projects/hslayers/src/components/print/models/text-style.model.ts (hunk 2 above)
Comment (DailisLangovskis) [reply projects/hslayers/src/components/print/models/text-style.model.ts:9]: textDraw I use for setting text title style drawn.

**Review by raitisbe — CHANGES_REQUESTED** [review]
A duplicate scale control gets added to the Ol map when printing once with Scale line enabled, then closing the new image tab and changing scaleline options.

**Comment by DailisLangovskis** [comment]
> A duplicate scale control gets added to the Ol map when printing once with Scale line enabled, then closing the new image tab and changing scaleline options.
> 
> image

This was not a problem few days ago, but will try to find the issue :D

**Review by FilipLeitner — APPROVED** [review]
Really nice :) but as raitis menitoned the landscape orientation should be default

**Comment by DailisLangovskis** [comment]
> 

It is implemented, just now specified landscape instead of size in cm.

**Pull request merged**